Prettify console output using Spectre.Console

ryandle opened the pull request:
This is the PR for issue #21 which updates the console output to use [Spectre.Console](https://spectreconsole.net/) for pretty formatting and colors.

Here's the before/after comparison  between v0.5.0 of the global tool and the changes in this PR:

![image](https://user-images.githubusercontent.com/578591/147833129-c14b0a2b-8c73-4138-b91a-645412cfd53d.png)

A walkthrough of the changes:

- New library dependency on Spectre.Console that is used to output the new table format + new stats format + updated message/rule/warning lines.
- All of the colors/formatting used is defined in `Configuration/Theme.cs`. The intent is that these could be defined by the yaml config in the future.
- Helper class `ConsoleOut` for writing message/rule/warnings to output.
- `LiveOutputTable` is a wrapper around a Spectre Live table which outputs GC stats. It uses a `System.Threading.Channel` to handle communication between the threads receiving GC events and the thread writing to console output.
  - Some of its APIs are async, which created some awkward spots in `Program.cs` where we are executing in a synchronous event handler.
  - The Heap stats timer now uses Task.Run instead of Timer (which doesn't support async.)
  - The HandleConsoleInput dedicated thread now also uses Task.Run since ThreadStart delegates don't support async.
- `PrintUtilities` has copies of most methods that format headers/rows. The original methods output specially aligned strings, and these new methods output lists of spectre marked-up strings that will be passed to the live table.
- I removed `GC#` from the start of the row string since it was a little redundant with the table header.

Some concerns I have and would like input on:

1. There are now two formatting libraries being used, Sharprompt & Spectre.Console - if that is a concern I think we could replace the Sharprompt usage with Spectre.Console (but the other way likely doesn't work as Sharprompt does cover all of the scenarios Spectre does.)
2. The colors are not currently configurable. I think it is possible to do that, so the question is do we need it up front or add it later? There is one small blocking issue that the colors used in the `BreakdownChart` in `Program.cs` are hard-coded. There is a code comment there explaining why (no way to go from string->Spectre.Color  for the BreakdownChart as far as I can tell.)
3. Redirecting console output (e.g. `dotnet-gcmon -n <process> > out.txt`) doesn't work after this change. I'm not sure if it can be supported with spectre live table. I left the original `PrintUtilities` methods around so that we could special handle the case where console output is being redirected and fall back to the plaintext formatting. Should I go ahead and do that?
4. Since I'm using the Async APIs on the Channel, there were a few awkward sync-meets-async places in the code outlined above; should I try to switch to using the blocking Try* APIs on Channel instead, which should remove the async-related changes?

Happy new year!

### Outstanding Work

- [x] Finish light theme support
- [x] More unit tests to cover yaml theme config
- [x] Support console redirection
- [x] Update dcumentation

ryandle committed:
commiting prototype

ryandle committed:
Initial cleanup

ryandle committed:
Merge branch 'main' into spectre

ryandle committed:
fix spacing + bad merge

ryandle committed:
Cleanup

ryandle committed:
formatting cleanup

ryandle committed:
take latest stable spectre.console

MokoSan commented on src/GCRealTimeMon/Configuration/Theme.cs:
```diff
@@ -0,0 +1,45 @@
+
+        // Future - these could be configurable via yaml
+        // Color reference: https://spectreconsole.net/appendix/colors
+        internal class Constants
```
NIT: But can this class be made a ``static`` class?

ryandle replied:
Yes!

MokoSan commented:
>> There are now two formatting libraries being used, Sharprompt & Spectre.Console - if that is a concern I think we could replace the Sharprompt usage with Spectre.Console (but the other way likely doesn't work as Sharprompt does cover all of the scenarios Spectre does.)

Totally agreed - I was experimenting with the Spectre.Console API and it seems like it'll be easy to incorporate. I can start working on a PR to facilitate this once this PR is checked in. 

This change looks AMAZING! Looking forward to seeing this in action.

MokoSan commented on src/GCRealTimeMon/Utilities/PrintUtilities.cs:
```diff
@@ -20,12 +22,34 @@ public static string GetHeader(Configuration.Configuration configuration)
+            // Iterate through all columns in the config.
+            foreach (var columnName in configuration.Columns)
+            {
+                //headerColumns.Add($"{FormatBasedOnColumnName(columnName)}");
```
NIT: We can probably get rid of this dead code.

ryandle committed:
cleanup based on review comments

Maoni0 commented:
thanks so much for the PR! 

I do think the colors would need to be configurable - when I run this in a cmd window with white background, it's not very readable -

![image](https://user-images.githubusercontent.com/10837357/148006848-84e04bac-dfb7-4e7b-8f5c-007c79f00cdb.png)

you could just have 2 defaults though - one for white bg and the other for dark bg. and if you use any other colored bg, you can config it yourself. 

> There are now two formatting libraries being used, Sharprompt & Spectre.Console - if that is a concern I think we could replace the Sharprompt usage with Spectre.Console (but the other way likely doesn't work as Sharprompt does cover all of the scenarios Spectre does.)

I'm usually for having less components. I take it that Sharprompt does not do colors? 

> Redirecting console output (e.g. dotnet-gcmon -n <process> > out.txt) doesn't work after this change. 

it's definitely convenient to have the console output redirectable. if spectre simply cannot be made work for this, what do you think about making the output configurable, ie, by default it uses spectre but if you want to redirect, use the old way?

> Since I'm using the Async APIs on the Channel, there were a few awkward sync-meets-async places in the code outlined above; should I try to switch to using the blocking Try* APIs on Channel instead, which should remove the async-related changes?

what kind of benefits do you believe async would provide here?

ryandle commented:
* RE: Color config - Alright, makes sense to me, I will make an update with configurable colors!
* RE: Sharprompt - I believe Sharprompt supports colors but it's only targted for scenarios where you prompt the user for input, i.e. it won't help with the pretty table output or the stats output. Spectre on the other hand does have prompt support too, so I think it's a good overall fit. @MokoSan was kind enough above to offer to replace Sharprompt usage with Spectre as a follow-up.
* RE: Console redirect - Supporting redirect also makes sense to me. I think that we can do something like what's happening #38 and detect the redirect case and automatically handle it? That way the user doesn't have to remember an extra flag.
* RE: Async apis on channel -None really, I just naturally didn't want to use a Try* pattern. After some reading, the expectation is that the Try* APIs on an unbounded channel would always return true, so I will switch to that and assert true.
    * There will still be a `StopAsync` method on `LiveOutputTable`. I don't see a great way to know it is stopped without waiting for the task returned by Spectre's `LiveDisplay.StartAsync`. Looking at the code behind `LiveDisplay`, it is doing the console rendering work in a separate task continuation, so in situations like where we need to print stats to the console, we have to wait for the LiveDisplay's task to complete to be sure that it is not still writing to the console.

ryandle commented:
Working on a light color theme and ran into a small issue with the breakdown chart used in the stats view. Opened this issue to investigate a resolution: https://github.com/spectreconsole/spectre.console/issues/684

ryandle committed:
- Break theme out into its own subfolder. Introduce light and dark theme defaults. Add CustomTheme that can be defined in yaml and falls back to the default theme when specific properties aren't defined. The default theme is dark, but switches to light if the console background color is White.
- Changed LiveOutputTable.WriteRowAsync to be synchronous
- Cleaned up use of wrong namespace and made it the default for the csproj
- Added two basic UTs; more to come;

ryandle committed:
add theme serialization UT

ryandle committed:
- Refactored capturing TraceGC event data into a single object so we can pass that reference around and not worry about tearing between the datetime & TraceGC data.
- Introduced a new interface for console output responsibilities.Factored the Spectre code into one implementation. Will move the plaintext code back into a 2nd implementation. We'll have a factory method choose the implementation at runtime based on whether output is redirected && config.
- Got rid of Restart in LiveOutput Table...Start/Stop should be enough. Added a check to prevent starting twice.

ryandle committed:
- Implemented PlainText console output (brings back the original implementation)
- Added Factory to decide output based on yaml config or console output being redirected
- Trimmed the IConsoleOuput interface to have only what is needed by Program.cs

ryandle committed:
Cleanup usings to match existing style

ryandle committed:
add documentation

ryandle commented:
Ok, @Maoni0 and @MokoSan I think this is ready for another pass. Here are some notes on the most recent changes:

* I changed the `WriteRowAsync` method on the `LiveOutputTable` to be synchronous. Start/Stop still have to be async because that is the API used by the Spectre.LiveTable.
* Theming is now fully configurable via yaml. The default is a dark theme, unless we detect Console.BackgroundColor to be white, then it's a light theme. I've updated the readme documentation with information about all of this.
* For the light theme, here is what I have so far, open to suggestions on better default color choices. (Also note the issue in Spectre I mentioned above about grey always being used in the bar char - that will remain for now until we get input form the Spectre maintainers.)
![image](https://user-images.githubusercontent.com/578591/148690708-a133d3e6-f4dc-47be-a65d-d072ba4ca12e.png)
* I added plain text support back. There is now an interface `IConsoleOutput` that has implementations for Spectre and plain text.
* You can enable plaintext either via the yaml config or by redirecting stdout. I started to add command line parameter support too. I was thinking `-nc` and `--no-color` - but the current commandline options library only supports single character 'short' option names, and since `p` and `n` are already taken I didn't have a good idea for a param character to represent 'no color'. Open to ideas!

MokoSan commented:
Looks great!! Just tried this out on Linux to make sure we are good there, as well and things look good:

![image](https://user-images.githubusercontent.com/4951960/148716543-bfce7ed6-7e20-4170-8088-75f8aa227166.png)

Details of the test machine:
```
MokoSan:GCRealTimeMon:% lsb_release -a                                                                                                                                                                
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04 LTS
Release:        20.04
Codename:       focal
```

One note I am adding for myself when I tackle https://github.com/Maoni0/realmon/issues/40 is to include a prompt on the configuration side of things to include the ability to change the themes if the user wants to reconfigure the default config. 

>> You can enable plaintext either via the yaml config or by redirecting stdout. I started to add command line parameter support too. I was thinking -nc and --no-color - but the current commandline options library only supports single character 'short' option names, and since p and n are already taken I didn't have a good idea for a param character to represent 'no color'. Open to ideas!

``-d`` => devoid of color? We can still keep ``--no-color`` as the long name? LOL, sorry, couldn't think of anything better. 😛

Maoni0 commented:
that's wonderful! I will take a look at the changes later today, thanks for making these changes, @ryandle!

Maoni0 commented on src/GCRealTimeMon/Configuration/Theme/CustomTheme.cs:
```diff
@@ -0,0 +1,125 @@
+            set => gen4HeapColor = value;
+        }
+
+        public string HighlightColor
```
why don't these fields have an alias?

ryandle replied:
In `ConfigurationReader` we're using `UnderscoredNamingConvention` to parse the yml config which turns camelCase property names into camel_case. That works fine for the properties below, but these I thought were awkward as the default convention would have been `g_c_header_color` or `gen_0_heap_color`.  BTW I put this at the top as a remark on the class, but I guess this Q shows its not a great place for the comment.

What do you think? Do you want to keep the default convention, or override it w/ these attributes?

Maoni0 replied:
oh I see. I forgot about the camelCase -> camel_case conversion 😀 I think this is fine. thanks for the explanation!

Maoni0 commented on README.md:
```diff
@@ -123,6 +123,29 @@ The process to add a new column from the ``TraceGC`` event is the following:
+| total_heap_color      | The color used to for information about total heap stats in the heap stats table.                                                                 |
+| highlight_color       | The color used to call out special information, like the process being monitored.                                                                 |
+| message_color         | The color used for standard messages written to output.                                                                                           |
+| warning_color         | The color used for warning messages written to output.                                                                                            |
```
these seem to all be for the stats table? what about colors for printing out the individual GC events?

ryandle replied:
The individual GC events are the "row_color" fields above this on lines 136-138.

Message/Warning color were not intended to be specific to the heap stats table but... I guess it ended up that way 😅The two other messages we write to output right now are "ruled" lines `-------- Message --------` which are currently covered by `MessageRuleColor` (which I forgot to document here!)

I think `MessageRuleColor` should be removed and we just use `MessageColor` for all general messages. The only warning right now is when there are no stats to print, but I think I'd lean towards keeping it as a general warning color config (not heap stats specific) for when other warnings are added. Let me know what you think!

Maoni0 replied:
yes, I'm for keeping MessageColor for all general messages.

Maoni0 commented:
btw, do you mind explaining what is a "Spectre.Console live table"? what makes it live? 😀 does it need to look to see if the color specification changes while the process is running?

Maoni0 commented on src/GCRealTimeMon/Utilities/PrintUtilities.cs:
```diff
@@ -48,7 +71,7 @@ public static string GetLineSeparator(Configuration.Configuration configuration)
 
             string lineSeparator = "";
-            for(int i = 0; i < repeatCount; i++)
+            for (int i = 0; i < repeatCount; i++)
```
> for (int i = 0; i < repeatCount; i++) [](http://example.com/codeflow?start=11&length=38)

nit - I guess your formatting is to have a space for the for/foreach statement, I do notice some inconsistency in the src - we should standize the formating. is there any conventional formatting, is this "space before '(' unless it's a method call" a convention in VS? if there's already something we use that and reformat our src.

ryandle replied:
VS has this setting which would apply to all control flow keywords not just for/foreach. Based on the source history I'm thinking you also have it set to enabled Maoni? I'm pretty sure it's enabled by default in VS:
![image](https://user-images.githubusercontent.com/578591/149262478-e311d99a-8fbd-46a7-b762-a184bcb4a10b.png)

@MokoSan this commit was the first I found of the "no space between control flow keywords" - is that your preference or perhaps it was a misconfiguration? https://github.com/Maoni0/realmon/commit/84080aaf4a051031b496bab09bbe97988692bc40 

(edit well now I see in that commit that some keywords like `if` have a space before parens and others like `foreach` don't - haven't found a style setting like that, wondering if it was just an unformatted doc?)

Maoni0 commented on src/GCRealTimeMon/Utilities/SpectreConsoleOut.cs:
```diff
@@ -0,0 +1,128 @@
+        /// <summary>
+        /// Writes a horizontal rule line to the console.
+        /// </summary>
+        public void WriteRule()
```
what's the definition of a rule?

ryandle replied:
It's a horizontal rule, like the `<hr/>` tag in html. 

I think it's short for ruler? I'm not quite sure, I was thinking of hr and rule as synonymous in this context. Now you have me wondering...

Edit: yeah I think it's just short for ruler aka straight line: https://en.wikipedia.org/wiki/Ruler

Maoni0 replied:
got it, thanks!

hez2010 commented:
> btw, do you mind explaining what is a "Spectre.Console live table"? what makes it live? 😀 does it need to look to see if the color specification changes while the process is running?

The "live" makes it able to update arbitrary widgets in-place, see https://spectreconsole.net/live/live-display

Maoni0 commented:
thanks @hez2010. from the video it seems like "live" means "I can overwrite what's already displayed" but that's not what we are doing here - we don't need to overwrite anything that's already displayed. or did I understand it wrong?

ryandle commented:
@Maoni0 I haven't yet gotten a block of time where I can sit down and reply to everything at once - but figured I should touch on the Live table Q.

So normal Spectre widgets are "Give it all the data at once and print it to the screen in your desired format."  So to use a normal Spectre table, you need to give it all data up front and a full table is rendered.

The scenario for this tool is different, we want to "stream" data to the console as GC events happen. That seemed like a good fit for the Live table - which is a table where you can keep appending rows over time - you don't have to have all the data up front.

I originally tried to use the non-live table widget, thinking I could just not write the table header every time there is a new GC event and that might effectively just append a row. It didn't work as well though - IIRC there were still things like spacing before/after each table (i.e. each GC event) that made the data a lot less dense than a single continuous table.

(I also considered not using a table at all and just doing more custom formatting - in the end I thought the table looked the best and also reduced how much custom formatting/layout code we would maintain.)

I hope this helps answer your Q. LMK if you have more! And I will work on adding these details to a code comment on LiveOutputTable so it's clear what is being used and why.

Maoni0 commented:
got it, thanks for the explanation, @ryandle! and no worries at all about responding - respond when it's convenient for you :)

ryandle committed:
Merge branch 'main' into spectre

ryandle committed:
fixing spacing after merge

ryandle committed:
fix typo

Maoni0 commented:
did you want to get rid of `MessageRuleColor` as we discussed above? other than that it looks great to me!

ryandle commented:
Hey @Maoni0, I'm pushing the change now. One thing I missed above is that in the dark theme the "rule" color was different from the message color, so you had a green ruler w/ blue message like this:

![image](https://user-images.githubusercontent.com/578591/150138385-a1ac5672-3fda-47dc-b3b7-f902d5ab500b.png)

And after the change to remove the rule color, we'll just have a single color for both like this:

![image](https://user-images.githubusercontent.com/578591/150138298-80a22e70-4bcc-4dda-aaed-d664f8e986b9.png)

Which is fine (I was using the same color for both already in the Light theme) I just wanted to call it out.

ryandle committed:
Remove Rule color, use MessageColor instead.

Maoni0 commented:
cool; we are good to go. merging now!
